drivers/serial/uart_ns16550.c

```diff
@@ -733,10 +821,13 @@ static void uart_ns16550_isr(void *arg)
 {
 	struct device *dev = arg;
 	struct uart_ns16550_dev_data_t * const dev_data = DEV_DATA(dev);
+	k_spinlock_key_t key = k_spin_lock(&dev_data->lock);
```

Comment:
Had some time to come back and look at this this morning.  Here is the deadlock: if the driver takes the lock before calling the registered handler, then the next uart action the handler does (for example: "try to read a byte") will deadlock, because the lock is already taken.

Just remove the locking from this spot, it's not needed.   Nothing is touching the hardware here, so there's nothing to protect.  Whatever the handler does already has its own locking.

Reply:
> yes, also my finding: [#23693 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/23693#discussion_r425886802)

This has been removed/resolved in the latest commit to #23693. Thanks for the feedback!